File: java/25. Reverse Nodes in k-Group/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        return reverseKGroup(head, k, countNodes(head));
    }

    public ListNode reverseKGroup(ListNode head, int k, int remaining) {
        if (remaining < k) return head;
        ListNode tail = kthNode(head, k - 1);
        ListNode p = tail.next;
        tail.next = null;
        ListNode nextHead = reverseKGroup(p, k, remaining - k);
        ListNode currentHead = reverse(head);
        head.next = nextHead;
        return currentHead;
    }


    public int countNodes(ListNode head) {
        if (head == null) return 0;
        ListNode current = head;
        int count = 1;
        while (current.next != null) {
            current = current.next;
            count++;
        }
        return count;
    }

    public ListNode kthNode(ListNode head, int k) {
        ListNode current = head;
        for (int i = 0; i < k; i++) {
            current = current.next;
        }
        return current;
    }

    public ListNode reverse(ListNode head) {
        if (head == null || head.next == null) return head;

        ListNode p = reverse(head.next);
        head.next.next = head;
        head.next = null;
        return p;
    }
}

File: python/25. Reverse Nodes in k-Group/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        # assume at least k items exist in the given linkedlist
        def reverse(head, k):
            prev = None
            current = head

            while k > 0:
                next = current.next
                current.next = prev
                prev = current
                current = next
                k -= 1

            return prev

        count = 0
        current = head

        # go k step forward
        while count < k and current:
            current = current.next
            count += 1

        if count == k:
            # we need to reverse k nodes
            newHead = reverse(head, k)
            # here head given to the reverse function will be
            # the tail of the reversed part of the linked list
            # we need to connect it with the next batch
            head.next = self.reverseKGroup(current, k)
            return newHead

        else:
            # we dont have enough node, return head without reversing
            return head
